### PR TITLE
e2e: Initial e2e code

### PR DESCRIPTION
**Pull request by dougsland**
```
$ ./run-test-e2e
[ INFO ] Starting setup
[ INFO ] ==============================
[ INFO ] Cleaning any previous e2e files
[ INFO ] Creating podman network [netcow]

[ INFO ] Preparing ASIL environment
[ INFO ] ==============================
[ INFO ] Creating container control [ASIL mode]
[ INFO ] Creating stub systemd service: container-safety in container control
[ INFO ] Creating stub systemd service: container-cruise_control in container control
[ INFO ] Creating stub systemd service: container-tires in container control
[ INFO ] Creating stub systemd service: container-breaks in container control

[ INFO ] Preparing QM environment
[ INFO ] ==============================
[ INFO ] Creating container node1 [QM mode]
[ INFO ] Creating containers radio, maps, store
[ INFO ] See: https://github.com/containers/qm/issues/32

[ INFO ] Starting tests
[ INFO ] ==============================
[ INFO ] Waiting hirte containers be ready...

[ INFO ] #1- Test scenario: hirte list all units from control to node (vise-versa)
[ INFO ] Connected to control, listing few systemd units from control
[ INFO ] Executing: podman exec control hirtectl list-units control | head -5
ID                                                                              |   ACTIVE|      SUB
====================================================================================================
dbus.socket                                                                     |   active|  running
dev-cpu-19-msr.mount                                                            |   active|  mounted
integritysetup.target                                                           |   active|   active

[ INFO ] Connected to control, listing few systemd units from node1
[ INFO ] Executing: podman exec control hirtectl list-units node1 | head -5
ID                                                                              |   ACTIVE|      SUB
====================================================================================================
sys-fs-fuse-connections.mount                                                   | inactive|     dead
systemd-machine-id-commit.service                                               | inactive|     dead
dev-cpu-4-cpuid.mount                                                           |   active|  mounted

[ INFO ] All set!
```

### CODE REVIEW

**Comment by dougsland** [comment]
`before merge`:
     - README what need to run the e2e
     - README what this e2e script executes and tests?
     - README: document the env variables that change the behaviour of script:`NUMBER_OF_NODES, TAG_CONTROL_MACHINE, NET_INTERFACE_IP_CONTROL, TAG_NODE_MACHINE`
     - SCRIPT: finish QM setup and trigger the containers (see QM issue #32)
     - cosmetic: shellcheck in the code
     - feedback from others devs

**Comment by lsm5** [comment]
Not a review, just a note: it would be great to integrate these tests with Packit and TMT (https://packit.dev/docs/testing-farm/) . Doesn't have to be in this PR, and fwiw, we don't have e2e testing with packit complete in any of our repos yet.

**Comment by dougsland** [comment]
> Not a review, just a note: it would be great to integrate these tests with Packit and TMT (https://packit.dev/docs/testing
-farm/) . Doesn't have to be in this PR, and fwiw, we don't have e2e testing with packit complete in any of our repos yet.

Thanks @lsm5. Indeed, it will require a different github issue (as soon we merge) as I will need to research the tool to start using it.

**Comment by rhatdan** [comment]
@dougsland Still a WIP?

**Comment by lsm5** [comment]
> > Not a review, just a note: it would be great to integrate these tests with Packit and TMT (https://packit.dev/docs/testing
> > -farm/) . Doesn't have to be in this PR, and fwiw, we don't have e2e testing with packit complete in any of our repos yet.
> 
> Thanks @lsm5. Indeed, it will require a different github issue (as soon we merge) as I will need to research the tool to start using it.

https://github.com/containers/container-selinux/pull/230 might be of some use. It's not merged yet and actually running those tests does cause `out of disk space` issues on the testing env, that I have yet to check on. TMT does allow specifying hardware configuration, but I haven't yet gotten it to work successfully: https://tmt.readthedocs.io/en/stable/spec/hardware.html#disk

**Comment by dougsland** [comment]
> @dougsland Still a WIP?

@rhatdan almost there, we need a README and a fresh test "tomorrow". As today I got surprised with some changes. 👍

**Review by lsm5 — CHANGES_REQUESTED** [review]
Could you please also add in the commit message how to actually run the test (which dir to run from), as root / non-root etc?

**Comment by lsm5** [comment]
> Could you please also add in the commit message how to actually run the test (which dir to run from), as root / non-root etc?

and also in the README

File: tests/e2e/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# e2e
+`run-test-e2e` executes tests for qm software using [podman quadlet](https://www.redhat.com/sysadmin/quadlet-podman) as platform.
+
+[![asciicast](https://asciinema.org/a/cwnb6RjckO7vXLUvHpbMA9fAU.svg)](https://asciinema.org/a/cwnb6RjckO7vXLUvHpbMA9fAU)
+
+## Requirements
+A recent version of Fedora or CentOS9 with the following packages: bash, selinux and podman. 
+Make sure to enable `cgroupv2`.
+
+## Tested executed
```
Comment (lsm5) [inline tests/e2e/README.md:17]: Nit: maybe `tested and executed` or `Tests executed` ?

File: tests/e2e/lib/container [hunk 2]
```diff
@@ -0,0 +1,152 @@
+#!/bin/bash
+# shellcheck disable=SC2129
+#
+# Copyright 2023 The qm Authors
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply tests/e2e/lib/container:4]: No need but no problem either.

**Comment by rhatdan** [comment]
Cleanup @lsm5 issues, and then we can merge.  We need to get this test to run on PRs.

**Comment by lsm5** [comment]
I think if we need something running quickly, github actions might be the way, although I don't think it's easy to provision anything other than Ubuntu over there. Packit + TMT might be the way to go long-term.

**Comment by dougsland** [comment]
> Could you please also add in the commit message how to actually run the test (which dir to run from), as root / non-root
> etc?

Just run ./run-e2e-test in the dir. There is a asciicinema embedded in the README that show as 'video' how to run it.

File: tests/e2e/lib/container (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougsland) [reply tests/e2e/lib/container:4]: IMO, tha'ts a good practice (not required). Many opensource projects do it since the early days.

**Comment by dougsland** [comment]
> Could you please also add in the commit message how to actually run the test (which dir to run from), as root / non-root etc?

The script will warn if user try to run as non root.

**Comment by dougsland** [comment]
@rhatdan  @lsm5 here how the HTML will look like when README.md is merged: https://github.com/dougsland/qm/tree/e2e/tests/e2e

**Pull request merged**